---
title: "Nematode clusters evaluation"
output:
  output: github_document
---

```{r, warning=FALSE, message=FALSE}
# Load resources. Libraries, paths, functions, themes.
source(file.path("..", "..", "scripts", "config.R"))
set.seed(101)
```

```{r}
# Read merged dataset
merge_at_nema <- readRDS(ATH_DATA_PATH)
merge_os_nema <- readRDS(OSA_DATA_PATH)
```

## Selection of nematode clusters by frequency of cells from infected samples

```{r}
#Umaps split by treatment

DimPlot(merge_at_nema, group.by = "Treatment", cols = cols_treat, 
        split.by = "Treatment")& 
  theme_void()& 
  ggtitle(NULL) &
  theme(legend.position = "none",
        strip.text.x = element_blank())
```
```{r}
#Umaps split by treatment

DimPlot(merge_os_nema, group.by = "Treatment", cols = cols_treat, 
        split.by = "Treatment")& 
  theme_void()& 
  ggtitle(NULL) &
  theme(legend.position = "none",
        strip.text.x = element_blank())
```


```{r}
# Stats of cluster frequency

treatment_nema <- merge_at_nema@meta.data %>%
  select(Cell_type, Treatment, Cluster)


test <- chisq.test(table(treatment_nema$Cluster, treatment_nema$Treatment))
df <- table(treatment_nema$Cluster, treatment_nema$Treatment)
df1 <- cbind(df, t(apply(df, 1, function(x) {
  ch <- chisq.test(x, p = c(0.716, 0.284))
  c(unname(ch$statistic), ch$p.value, ch$expected)
})))
colnames(df1) <- c("Mock_obs", "Inf_obs", "xsquared", "pvalue", "Mock_exp", "Inf_exp")
df1 <- df1 %>%
  as.data.frame() %>%
  mutate(
    freq = (Inf_obs / Mock_obs),
    freq_fold = (Inf_obs / Mock_obs) / 0.284
  ) %>%
  arrange(desc(freq_fold))
df1 %>% 
  head(5)
```

```{r}
# Sort cluster by frequency
merge_at_nema@meta.data <- merge_at_nema@meta.data %>%
  mutate(Cluster = fct_relevel(Cluster, rownames(df1)))
at_cluster_treat <- ggplot(data = merge_at_nema@meta.data, aes(x = Cluster, fill = factor(Treatment))) +
  geom_bar(position = "fill") +
  scale_fill_manual(values = cols_treat) +
  theme_classic() +
  geom_hline(yintercept = 0.284, linetype = "dashed", linewidth = 0.1) +
  ylab("Frequency") +
  theme_paper +
  xlab("Cluster") +
  coord_flip()

at_cluster_treat
```


```{r}
#Stats of cluster frequency

treatment_nema <- merge_os_nema@meta.data%>% 
  select(Treatment,Cluster)


test <- chisq.test(table(treatment_nema$Cluster,treatment_nema$Treatment))
df<-table(treatment_nema$Cluster,treatment_nema$Treatment)
df1 <- cbind(df, t(apply(df, 1, function(x) {
  ch <- chisq.test(x,p = c(0.378,0.622))
  c(unname(ch$statistic), ch$p.value,ch$expected)})))
colnames(df1) <- c("Mock_obs","Inf_obs",'xsquared', 'pvalue',"Mock_exp", "Inf_exp")
df1<-df1 %>% as.data.frame() %>% mutate(freq=(Inf_obs/Mock_obs),
                                        freq_fold=(Inf_obs/Mock_obs)/0.622) %>% 
  arrange(desc(freq_fold))
df1 %>% 
  head(5)
```

```{r}
#Sort cluster by frequency
merge_os_nema@meta.data<-merge_os_nema@meta.data %>% 
  mutate(Cluster=fct_relevel(Cluster, rownames(df1)))
os_cluster_treat <- ggplot(data=merge_os_nema@meta.data, aes(x=Cluster, fill=factor(Treatment))) +
  geom_bar(position="fill")+
  scale_fill_manual(values=cols_treat)+
  theme_classic()+
  geom_hline(yintercept=0.622,linetype = "dashed",linewidth=0.1)+
  ylab("Frequency")+
 theme_paper+
  xlab("Cluster")+
  coord_flip()

os_cluster_treat
```
## Gene expression of known RKN responsive genes 


```{r}

#FeaturePlots of known markers
#WRKY23
wrky23<-FeaturePlot(merge_at_nema,"AT2G47260",pt.size = 0.05,order=TRUE,keep.scale="all")+
  scale_colour_gradientn(colours=rev(brewer.pal(11,"PiYG")[1:6]))+ggtitle("AtWRKY23")+
  theme_void()
lox<-FeaturePlot(merge_at_nema,"AT1G17420",pt.size = 0.05,order=TRUE,keep.scale="all")+
 scale_colour_gradientn(colours=rev(brewer.pal(11,"PiYG")[1:6]))+ggtitle("AtLOX3")+
  theme_void()
pdf2.2<-FeaturePlot(merge_at_nema,"AT2G02100",pt.size = 0.05,order=TRUE,keep.scale="all")+
  scale_colour_gradientn(colours=rev(brewer.pal(11,"PiYG")[1:6]))+ggtitle("AtPDF2.2")+
  theme_void()
```

```{r, fig.width=12, fig.height=4}
wrky23 +lox+pdf2.2+plot_layout(ncol = 3)
```

```{r}
pdf <- FeaturePlot(merge_os_nema,"Os02g0629800",pt.size = 0.05,order=TRUE,keep.scale="all")+
  scale_colour_gradientn(colours=rev(brewer.pal(11,"PiYG")[1:6]))+
  theme_void()+ggtitle("osDEF7")

del1 <- FeaturePlot(merge_os_nema,"Os02g0739700",pt.size = 0.05,order=TRUE,keep.scale="all")+
 scale_colour_gradientn(colours=rev(brewer.pal(11,"PiYG")[1:6]))+
  theme_void()+ggtitle("osDEL1")

erf <- FeaturePlot(merge_os_nema,"Os09g0457900",pt.size = 0.05,order=TRUE,keep.scale="all")+
  scale_colour_gradientn(colours=rev(brewer.pal(11,"PiYG")[1:6]))+
  theme_void()+ggtitle("OsERF102")
```

```{r, fig.width=12, fig.height=4}
pdf +del1+erf+plot_layout(ncol = 3)
```
## Nematode score calculation

We are going to use bulK RNA-seq studies under nematode infection to evaluate in our scRNA-seq dataset which cells are showing a profile that corresponds to nematode response.

### Nematode score in *Arabidopsis* dataset

```{r}
# Directory resources

directory_resources <- file.path("..","..","data","resources","03_Nematode_clusters")
```


```{r}
#Read files upregulated genes and create matrix
files_up <- list.files(path=(file.path(directory_resources,"Ath_up")),full.names = TRUE)
up_data <- files_up %>%
  map_dfr(~read.csv(.x, header = TRUE) %>% 
            select(Gene) %>%  # Only keep the Gene column
            mutate(file = tools::file_path_sans_ext(basename(.x))),  # Add file name without extension
          .id = "file_id")
# Create the presence/absence matrix
gene_matrix <- up_data %>%
  mutate(present = 1) %>%  # Add a column to indicate presence
  pivot_wider(
    id_cols = Gene,
    names_from = file,
    values_from = present,
    values_fill = 0  # Fill NA values with 0
  ) %>%
  column_to_rownames("Gene")  # Make Gene the row names

gene_matrix <- gene_matrix %>% mutate(across(everything(), as.numeric))

head(gene_matrix)
```

```{r}
#write.table(gene_matrix, file = file.path(directory_resources,"ath_gene_presence_matrix.txt"), quote = FALSE, row.names = TRUE, col.names = TRUE, sep = "\t")
```

```{r}
# Calculate row sums
up_data_sum <- gene_matrix %>%
  as.data.frame() %>%
  rownames_to_column("Gene") %>%
  dplyr::mutate(sum = rowSums(across(where(is.numeric))))
```

```{r}
ath_bulk_bar <- ggplot(up_data_sum, aes(factor(sum),fill=factor(sum))) +
  geom_bar(stat="count", position = "dodge")+
  scale_fill_manual(values = c("grey90","grey90",
                              "#DE77AE","#DE77AE","#DE77AE"))+
  theme_classic()+
  coord_flip()+
  ylab("Up-regulated genes")+
  xlab(NULL)+
  scale_y_continuous(limits = c(0, 1250), breaks = c(0, 500, 1000))+
  theme_notebook

```

```{r}
ath_bulk_bar
```


```{r}
#Select genes upregulated in at least 3 studies
nema_up_core <- gene_matrix[rowSums(gene_matrix,na.rm=TRUE)>2,]
nema_up_core <- rownames(nema_up_core)
```

```{r}
#write.table(nema_up_core, file = file.path(directory_resources,"ath_nema_core_genes.txt"), quote = FALSE, row.names = FALSE, col.names = FALSE, sep = "\t")
```


```{r}
#Extract expression data of nematode core genes and calculate nematode score
count.data <- GetAssayData(merge_at_nema, slot="counts",assay = "SCT")

nema_up_data <- count.data[rownames(count.data) %in% nema_up_core,]
nema_up_data <- as.matrix(nema_up_data)
nema_up_data_z <-t(scale(t(nema_up_data)))
nema_sum_z <- colSums(nema_up_data_z)
nema_sum_z <- rescale(nema_sum_z,to=c(-1,5))
```

```{r}
#Assign nematode score and positive cells
merge_at_nema[["nema_score"]]<-nema_sum_z
nemacells <- WhichCells(merge_at_nema, expression = nema_score >0.5)
merge_at_nema$nema_profile<- ifelse(colnames(merge_at_nema) %in% nemacells, "Pos", "Neg")
```

```{r}
ath_nema_score_umap <- DimPlot(merge_at_nema, group.by = "nema_profile",cols = c("grey90","#DE77AE"),order = "Pos",pt.size = 0.05)+
  theme_void()+
  theme(legend.position="none")+
  ggtitle(NULL)
```


```{r}
ath_nema_score_umap
```


```{r}
#Proportion of cells from mock and infected samples assigned as positive for nematode responsive cells
table(merge_at_nema@meta.data$nema_profile,merge_at_nema@meta.data$Treatment) %>% prop.table(2)
```

```{r,warning=FALSE}
density_nema <- merge_at_nema@meta.data %>% 
  select(Treatment,nema_score,nema_profile,Nema_cluster) %>% 
  mutate(Nema_cluster = factor(Nema_cluster, levels = c("Other","At8","At28")))

density_score <- ggplot(density_nema, aes(x=nema_score, fill=Nema_cluster)) +
  annotate("rect", xmin = 0.5, xmax =3, ymin = 0, ymax = 2, 
           alpha = 0.6,fill="#DE77AE")+
  geom_density(alpha=0.8,size=0.1)+
  scale_x_continuous(limits = c(-1, 3), breaks = c(-1,0.5,3))+
  scale_y_continuous(limits = c(0, 2), breaks = c(0,1,2))+
  theme_notebook+
  scale_fill_manual(values=cols_nema_at)+
  geom_vline(xintercept = 0.5,
             linetype = 2,
             size=0.1,
             color = "#DE77AE")+
  ylab("Density")+
  xlab("Nematode score")+theme(axis.text.y=element_blank(),
                               axis.ticks.y=element_blank(),
                               panel.spacing = unit(0.2, "lines"),
                               strip.background = element_blank())+
  facet_grid(rows = vars(Nema_cluster))
```

```{r}
density_score
```

### Nematode score in rice dataset

```{r}
#Read files upregulated genes and create matrix
files_up <- list.files(path=(file.path(directory_resources,"Osa_up")),full.names = TRUE)
up_data <- files_up %>%
  map_dfr(~read.csv(.x, header = TRUE) %>% 
            select(Gene) %>%  # Only keep the Gene column
            mutate(file = tools::file_path_sans_ext(basename(.x))),  # Add file name without extension
          .id = "file_id")
# Create the presence/absence matrix
gene_matrix <- up_data %>%
  dplyr::group_by(Gene,file) %>% distinct() %>% #delete duplicated genes in a file
  ungroup() %>% 
  mutate(present = 1) %>%  # Add a column to indicate presence
  pivot_wider(
    id_cols = Gene,
    names_from = file,
    values_from = present,
    values_fill = 0  # Fill NA values with 0
  ) %>%
  column_to_rownames("Gene")  # Make Gene the row names

gene_matrix <- gene_matrix %>% mutate(across(everything(), as.numeric))

head(gene_matrix)
```


```{r, eval=FALSE}
#write.table(gene_matrix, file = file.path(directory_resources,"osa_gene_presence_matrix.txt"), quote = FALSE, row.names = TRUE, col.names = TRUE, sep = "\t")
```

```{r}
# Calculate row sums
up_data_sum <- gene_matrix %>%
  as.data.frame() %>%
  rownames_to_column("Gene") %>%
  dplyr::mutate(sum = rowSums(across(where(is.numeric))))
```

```{r}
osa_bulk_bar <- ggplot(up_data_sum, aes(factor(sum),fill=factor(sum))) +
  geom_bar(stat="count", position = "dodge")+
  scale_fill_manual(values = c("grey90","grey90",
                              "#DE77AE","#DE77AE","#DE77AE"))+
  theme_classic()+
  coord_flip()+
  ylab("Up-regulated genes")+
  xlab(NULL)+
  theme_notebook+
  scale_y_continuous(limits = c(0, 7000), breaks = c(0, 3500, 7000))

```

```{r}
osa_bulk_bar
```

```{r}
#Select genes upregulated in at least 3 studies
nema_up_core <- gene_matrix[rowSums(gene_matrix,na.rm=TRUE)>2,]
nema_up_core <- rownames(nema_up_core)
```

```{r}
#write.table(nema_up_core, file = file.path(directory_resources,"Nema_score","osa_nema_core_genes.txt"), quote = FALSE, row.names = FALSE, col.names = FALSE, sep = "\t")
```


```{r}
#Extract expression data of nematode core genes and calculate nematode score
count.data <- GetAssayData(merge_os_nema, slot="counts",assay = "SCT")

nema_up_data <- count.data[rownames(count.data) %in% nema_up_core,]
nema_up_data <- as.matrix(nema_up_data)
nema_up_data_z <-t(scale(t(nema_up_data)))
nema_sum_z <- colSums(nema_up_data_z)
nema_sum_z <- rescale(nema_sum_z,to=c(-1,5))
```

```{r}
#Assign nematode score and positive cells
merge_os_nema[["nema_score"]]<-nema_sum_z
nemacells <- WhichCells(merge_os_nema, expression = nema_score >1)
merge_os_nema$nema_profile<- ifelse(colnames(merge_os_nema) %in% nemacells, "Pos", "Neg")
```


```{r}
#Proportion of cells from mock and infected samples assigned as positive for nematode responsive cells
table(merge_os_nema@meta.data$nema_profile,merge_os_nema@meta.data$Treatment) %>% prop.table(2)
```

```{r}
osa_nema_score_umap <- DimPlot(merge_os_nema, group.by = "nema_profile",cols = c("grey90","#DE77AE"),order = "Pos",pt.size = 0.05)+
  theme_void()+
  theme(legend.position="none")+
  ggtitle(NULL)

```


```{r}
osa_nema_score_umap
```

```{r}
density_nema <- merge_os_nema@meta.data %>% 
  mutate(Nema_cluster = factor(Nema_cluster, levels = c("Other","Os3","Os9","Os10","Os16","Os17")))

density_score <- ggplot(density_nema, aes(x=nema_score, fill=Nema_cluster)) +
  annotate("rect", xmin = 1, xmax =2, ymin = 0, ymax = 1, 
           alpha = 0.6,fill="#DE77AE")+
  geom_density(alpha=0.8,size=0.1)+
  scale_x_continuous(limits = c(-1.1, 2), breaks = c(-1,1,2))+
  scale_y_continuous(limits = c(0, 1))+
  theme_notebook+
  scale_fill_manual(values=cols_nema_os)+
  geom_vline(xintercept = 1,
             linetype = 2,
             size=0.1,
             color = "#DE77AE")+
  ylab("Density")+
  xlab("Nematode score")+theme(axis.text.y=element_blank(),
                               axis.ticks.y=element_blank(),
                               panel.spacing = unit(0.1, "lines"),
                               strip.background = element_blank())+
  facet_grid(rows = vars(Nema_cluster))
```

```{r}
density_score
```

```{r}
sessionInfo()
```

